File: sorts/quick_sort_hoare_method.rb
class Solution
  def quicksort(nums, left, right)
    if left < right
      # Find pivot index
      pi = partition(nums, left, right)
      quicksort(nums, left, pi)
      quicksort(nums, pi + 1, right)
    end
    nums
  end

  def partition(nums, left, right)
    # Choose middle element as pivot
    pivot = nums[(left + right) / 2]
    i = left - 1
    j = right + 1

    loop do
      loop do
        i += 1
        break if nums[i] >= pivot
      end

      loop do
        j -= 1
        break if nums[j] <= pivot
      end

      break if i >= j

      # Swap elements
      nums[i], nums[j] = nums[j], nums[i]
    end

    j
  end

  def sort_array(nums)
    return nums if nums.empty?
    quicksort(nums, 0, nums.length - 1)
  end
end

File: sorts/quick_sort_hoare_method.py
# quick sort
class Solution:
    def quicksort(self, nums, left, right):
        if left < right:
            # pivot index
            pi = self.partition(nums, left, right)

        quicksort(nums, left, pi)
        quicksort(nums, pi + 1, right)

    def partition(self, nums, left, right):
        pivot = nums[(left + right) // 2]  # setting it to middle element
        i = left - 1
        j = right + 1

        while True:
            while True:
                i += 1
                if nums[i] >= pivot:
                    break
                
            while True:
                j -= 1
                if nums[j] <= pivot:
                    break

            if i >= j:
                return j

            nums[i], nums[j] = nums[j], nums[i]

    def sortArray(self, nums: List[int]) -> List[int]:
        self.quicksort(nums, min(nums), max(nums))